### DISCUSSION TITLE
TextArea seems to crash under special circumstances

### DISCUSSION THREAD

**Discussion by UlrichGoebel** [Textualize/textual#4561] [Q&A] [answered]
Hi,

to start with: I have found a workaround for the following problem - see below. So it might be helpful for others or just interesting to explain the reason for it. I can't do the latter.

In my application I need to define my own `__init__` in the `Main(App)` class. So first I make my stuff and then I call the `__init__` from the `super` class.

All that works fine for `Input` widgets, but crashes using `TextArea` widgets. I didn't test it for other widgets as `Select` or `RadioSet`.

Here a minimal example. In the real world the `buildWidgets` does much more of course.

```python
from textual.app import App
from textual.widgets import Input, TextArea

class Main(App):
  
    def __init__(self, driver_class=None, css_path=None, watch_css=False):
        self.buildWidgets()
        super().__init__(driver_class, css_path, watch_css)
            
    def buildWidgets(self):
        self.MyWidget = Input()                  # works
        # self.MyWidget = TextArea()            # doesn' work
            
    def compose(self):
        yield self.MyWidget
    
if __name__ == "__main__":
    app = Main()
    app.run()
```

Running this with `Input()` in `buildWidgets` works, using `TextArea()` doesn't. It ends with a really long Error Message:

```python
Traceback (most recent call last):
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/message_pump.py", line 209, in app
    return active_app.get()
LookupError: <ContextVar name='active_app' at 0x7f3a88919120>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ulrich/PythonHobby/CLUI/Textual/./Test.py", line 20, in <module>
    app = Main()
  File "/home/ulrich/PythonHobby/CLUI/Textual/./Test.py", line 9, in __init__
    self.buildWidgets()
  File "/home/ulrich/PythonHobby/CLUI/Textual/./Test.py", line 14, in buildWidgets
    self.MyWidget = TextArea()
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/widgets/_text_area.py", line 437, in __init__
    self._set_document(text, language)
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/widgets/_text_area.py", line 853, in _set_document
    self.move_cursor((0, 0))
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/widgets/_text_area.py", line 1586, in move_cursor
    self.selection = Selection.cursor(location)
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/reactive.py", line 276, in __set__
    self._check_watchers(obj, name, current_value)
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/reactive.py", line 301, in _check_watchers
    invoke_watcher(obj, private_watch_function, old_value, value)
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/reactive.py", line 77, in invoke_watcher
    watch_result = cast(WatchCallbackBothValuesType, watch_function)(
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/widgets/_text_area.py", line 583, in _watch_selection
    self.app.cursor_position = self.cursor_screen_offset
  File "/home/ulrich/PythonHobby/CLUI/Textual/.venv/lib/python3.10/site-packages/textual/message_pump.py", line 216, in app
    raise NoActiveAppError()
textual._context.NoActiveAppError
```

It seems clear that building the `TextArea` makes problems.

## Found Workaround?

The last line `textual._context.NoActiveAppError` gave me the idea to change the order in the `__init__`: first the `super().__init__` then my stuff. Then it works, even for `TextArea`. No idea why, so what... The order doesn't matter to me as long as I don't manipulate the parameters `driver_class`, `css_path` and `watch_css`. I can live with that.

Best regards
Ulrich

**Comment by davep** [comment]
Building widgets before your application has had the chance to initialise itself isn't generally going to be a good idea. At the very least you'd want to be making the `super` call *before* building the widgets.

**Reply by UlrichGoebel** [reply]
Thanks for the quick reply!
Is there a better place to build widgets? May be at the beginning of the `compose`, so it would come to something like that:

```python
class Main(App):
  
    def __init__(self, driver_class=None, css_path=None, watch_css=False):
        super().__init__(driver_class, css_path, watch_css)
        self.buildWidgets()
            
    def buildWidgets(self):
        self.MyWidget = TextArea()
            
    def compose(self):
        self.buildWidgets()
        yield self.MyWidget
    
if __name__ == "__main__":
    app = Main()
    app.run()
```

Or is there a suitable event? (`Load`?) So it would come to something like

```python
class Main():

    ...

    def on_something(self):
        self.buildWidgets()
```

**Reply by davep** [reply, marked as answer]
Within `compose` would be more common. In most code you'll find that a `compose` method simply yields the widgets into the DOM, and then they are [queried back](https://textual.textualize.io/guide/queries/) when they're needed. See [this dialog](https://github.com/davep/tinboard/blob/99012c2e84cbe0331e94a4466ad394f9908108b3/tinboard/screens/bookmark_input.py#L115-L140) in one of my own applications, for example.

Note by the way in your example code that you're calling `buildWidgets` twice, which means you're going to be building one set and then never using them, instead using the subsequently-built set.

You should find that [the tutorial](https://textual.textualize.io/tutorial/) introduces these ideas, and [the guide](https://textual.textualize.io/guide/) goes on to build upon them. In particular the [guide to App basics](https://textual.textualize.io/guide/app/), and especially it's [introduction of widgets and how to compose them](https://textual.textualize.io/guide/app/#widgets) should contain some useful background.

**Reply by UlrichGoebel** [reply]
Thanks for the links. What I intend to do is create UIs on an existing PostgreSQL databases with about 300 widgets, organized in about 25 forms. These forms should be instances of a class `Form` which I have to build. So perheps I will open a new topic about "organizing large numbers of textual widgets in forms". Maybe there are hints or even interest in such a project.

**Comment by jdx** [comment]
A fix for the native extension header issue has been submitted:

https://github.com/jdx/ruby/pull/1

This PR:
1. Copies headers from portable dependencies (openssl, libyaml, etc.) into the Ruby tarball's `include/` directory during packaging
2. Updates `rbconfig.rb` to include `-I#{prefix}/include` in CPPFLAGS so mkmf can find the headers when building native gems
3. Adds CI tests for `openssl` and `psych` gem installation on macOS and Linux

*Note: This comment was AI-generated using Claude Code*
